File: String/16_isValid.java
class Solution(){

    private HashMap<Character, Character> mappings;

    public Solution(){
        this.mappings = new HashMap<Character, Character>();
        this.mappings.put(')', '(');
        this.mappings.put(']', '[');
        this.mappings.put('}', '{');
    }

    public boolean isValid(String s){

        // initialize a stack
        Stack <Character> stack = new Stack<Character>;
        
        for (int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            // if the current character is a closing bracket
            if (this.mappings.containKey(c)){
                char topElement = stack.empty() ? '#' : stack.pop()

                if (topElement != this.mappings.get(c)){
                    return false;
                }
            } else {
                stack.push(c);
            }
        }
        return stack.isEmpty();
    }
}

File: 2-String/16_isValid.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        
        stack = []

        mapping = {")":"(", "]":"[", "}":"{"}
        # mapping = {"(":")", "[":"]", "{":"}"} # 错
        
        for char in s:

            # If the character is a closing bracket
            if char in mapping:
                # Pop the topmost element from the stack, if it is non empty
                # Otherwise assign a dummy value of '#' to the top_element variable
                top_element = stack.pop() if stack else '#'
                # The mapping for the opening bracker in our hash and the top
                # element of the stack don't match, return False
                if mapping[char] != top_element:
                    return False
            else:
                # We have an opening bracker, simply push it onto the stack
                stack.append(char)
            
            # In the end, if the stack is empty, then we have a valid expression
        return not stack

sol = Solution()
s = "(()){}[]"
ans = sol.isValid(s)
print(ans)